### PR TITLE
Mailchimp source and new destination

### LINKED ISSUE #790: Feature Request: Mailchimp Member Updates without email addresses
Plan:
- source that talks mailchimp to import generated mailchimpId and another other merge var
- that allows it to make to an internal id
- new destination for mailchimp that uses mailchimpId to update existing users.

### PR DESCRIPTION
**Pull request by bleonard**
Executing this plan: https://github.com/grouparoo/grouparoo/issues/790

- source that talks mailchimp to import generated mailchimpId and another other merge var
- that allows it to make to an internal id
- new destination for mailchimp that uses mailchimpId to update existing users.

### CODE REVIEW

File: plugins/@grouparoo/marketo/__tests__/integration/destination-mapping-options.ts [hunk 1]
```diff
@@ -61,6 +61,6 @@ describe("marketo/destinationMappingOptions", () => {
       expect(urgency.type).toBe("float");
       expect(urgency.important).toBe(false);
     },
-    30 * 1000
+    1000 * 30
```
Comment (evantahler) [inline plugins/@grouparoo/marketo/__tests__/integration/destination-mapping-options.ts]: should be `helper.setupTime` to match other tests

File: plugins/@grouparoo/spec-helper/src/lib/specHelper.ts [hunk 2]
```diff
@@ -106,6 +106,8 @@ const models = [
 ];
 
 export namespace helper {
+  export const setupTime = 60 * 1000;
+
```
Comment (evantahler) [inline plugins/@grouparoo/spec-helper/src/lib/specHelper.ts]: That's a real shame.  If this is a problem just on CircleCI we can detect that with `process.env. CIRCLECI`.  I *think* there is value in keeping the timeout shorter normally to more quickly detect boot problems maybe?

File: plugins/@grouparoo/mailchimp/__tests__/data/profiles.csv [hunk 3]
```diff
@@ -0,0 +1,11 @@
+EMAIL,FNAME,LNAME,UserID,LTV
+xdiston9@grouparoo.com,Dorthy,Iston,10,290.53
```
Comment (evantahler) [inline plugins/@grouparoo/mailchimp/__tests__/data/profiles.csv:2]: lol, why `X`s?

File: plugins/@grouparoo/mailchimp/__tests__/import/profiles.ts [hunk 4]
```diff
@@ -0,0 +1,243 @@
...
+  SimpleDestinationOptions,
+} from "@grouparoo/core";
+
+const nockFile = path.join(__dirname, "../", "fixtures", "import-profiles.js");
+
+import { profiles } from "../../src/lib/import/profiles";
+
+// these comments to use nock
+const newNock = false;
+require("./../fixtures/import-profiles");
```
Comment (evantahler) [inline plugins/@grouparoo/mailchimp/__tests__/import/profiles.ts:27]: FYI I've been changing these to be `require (nockFile)`

File: plugins/@grouparoo/mailchimp/__tests__/import/profiles.ts [hunk 5]
```diff
@@ -0,0 +1,243 @@
...
+    importsCount,
+    sourceOffset: nextSourceOffset,
+  };
+}
+
+describe("mailchimp/import/profiles", () => {
+  beforeAll(async () => {
+    const env = await helper.prepareForAPITest();
+    actionhero = env.actionhero;
+  }, 1000 * 60);
```
Comment (evantahler) [inline plugins/@grouparoo/mailchimp/__tests__/import/profiles.ts]: `helper.timeout` probably?

File: plugins/@grouparoo/mailchimp/__tests__/integration/mailchimp-export-id.ts [hunk 6]
```diff
@@ -0,0 +1,350 @@
...
+    try {
+      const mailchimpId = generateMailchimpId(email);
+      await client.delete(`/lists/${listId}/members/${mailchimpId}`);
+    } catch (err) {}
+  }
+}
+
+async function cleanUp(suppressErrors) {
+  await deleteUsers(suppressErrors);
+}
```
Comment (evantahler) [inline plugins/@grouparoo/mailchimp/__tests__/integration/mailchimp-export-id.ts:81]: `suppressErrors` isn't used?

File: plugins/@grouparoo/mailchimp/__tests__/integration/mailchimp-import.ts [hunk 7]
```diff
@@ -0,0 +1,446 @@
...
+        session = await specHelper.buildConnection();
+        session.params = { email: "mario@example.com", password: "P@ssw0rd!" };
+        const sessionResponse = await specHelper.runAction(
+          "session:create",
+          session
+        );
+        expect(sessionResponse.error).toBeUndefined();
+        csrfToken = sessionResponse.csrfToken;
+
+        // create a sheet app with an uploaded file
```
Comment (evantahler) [inline plugins/@grouparoo/mailchimp/__tests__/integration/mailchimp-import.ts]: // mailchimp app

Comment (evantahler) [reply plugins/@grouparoo/mailchimp/__tests__/integration/mailchimp-import.ts]: There are a couple places in this file that say `sheet`

File: plugins/@grouparoo/mailchimp/src/lib/import/profileProperty.ts [hunk 8]
```diff
@@ -0,0 +1,7 @@
+import { ProfilePropertyPluginMethod } from "@grouparoo/core";
+
+export const profileProperty: ProfilePropertyPluginMethod = async () => {
+  // we always do nothing when called by profile update... we only exist to be used by the schedule & runs
+  // TODO: should this be made optional in the plugin definition?
```
Comment (evantahler) [inline plugins/@grouparoo/mailchimp/src/lib/import/profileProperty.ts:5]: https://grouparoo.height.app/general/T-707

File: core/api/src/tasks/schedule/run.ts [hunk 9]
```diff
@@ -29,6 +30,10 @@ export class ScheduleRun extends RetryableTask {
     await run.afterBatch();
 
     if (importsCount > 0) {
+      if (process.env.NODE_ENV === "test") {
+        // prevents "job already enqueued" in tests
+        params.count = (params.count || 0) + 1;
+      }
```
Comment (evantahler) [inline core/api/src/tasks/schedule/run.ts]: Which test was running into this error?  There's `specHelper.deleteEnqueuedTasks()` from `actionhero` which might be helpful

Comment (evantahler) [reply core/api/src/tasks/schedule/run.ts]: https://docs.actionherojs.com/modules/spechelper.html#deleteenqueuedtasks

**Review by evantahler — COMMENTED** [review]
Lots to like here! 
* Thanks for fixing up the test timeouts
* I think this pattern of adding new Connections for different source/destination behavior is the right thing to do (we'll make it better in the UI one day)

Things to talk about 
* ~~What's the migration path look like for existing `mailchimp-export` destination connection users look like?~~ All good! no name change

File: plugins/@grouparoo/mailchimp/__tests__/data/profiles.csv (hunk 3 above)
Comment (bleonard) [reply plugins/@grouparoo/mailchimp/__tests__/data/profiles.csv:2]: mailchimp remembers old ones i deleted, so I burned all the original ones in an early iterations 🤷

File: plugins/@grouparoo/mailchimp/__tests__/import/profiles.ts (hunk 4 above)
Comment (bleonard) [reply plugins/@grouparoo/mailchimp/__tests__/import/profiles.ts:27]: does that work? I thought I saw issues there originally.
Like it was run in some pre-compile step or something.

Comment (evantahler) [reply plugins/@grouparoo/mailchimp/__tests__/import/profiles.ts:27]: seems OK here? https://github.com/grouparoo/grouparoo/blob/master/plugins/%40grouparoo/bigquery/__tests__/table-import/import-properties.ts#L19-L28

File: core/api/src/tasks/schedule/run.ts (hunk 9 above)
Comment (bleonard) [reply core/api/src/tasks/schedule/run.ts]: the ones that looks this seems to be a risk
```
        // check that the run is enqueued
        const found = await specHelper.findEnqueuedTasks("schedule:run");
        expect(found.length).toEqual(1);
        expect(found[0].args[0].scheduleGuid).toBe(schedule.guid);

        // run the schedule
        const run = await Run.create({
          creatorGuid: schedule.guid,
          creatorType: "schedule",
          state: "running",
        });
        await specHelper.runTask("schedule:run", {
          runGuid: run.guid,
          scheduleGuid: schedule.guid,
        });
```

If it re-enqueues itself and falls on the same millisecond.
Maybe that's why there are these sleeps in the mysql one
```
      await helper.sleep();
      let foundAgain = await specHelper.findEnqueuedTasks("schedule:run");
      expect(foundAgain.length).toEqual(2);
      await specHelper.runTask("schedule:run", foundAgain[1].args[0]);

      await helper.sleep();
      foundAgain = await specHelper.findEnqueuedTasks("schedule:run");
      expect(foundAgain.length).toEqual(3);
      await specHelper.runTask("schedule:run", foundAgain[1].args[0]);
```

Comment (bleonard) [reply core/api/src/tasks/schedule/run.ts]: I'll remove the change to the task and add sleeps.

**Pull request merged**